File: backend/abonos.py
from datetime import date
from decimal import Decimal
from pathlib import Path

from backend.db import transaccion


CENTAVO = Decimal("0.01")


def dinero(valor: int | float | str | Decimal) -> Decimal:
    return Decimal(str(valor)).quantize(CENTAVO)


def registrar_abono(
    ruta_bd: str | Path,
    fecha: str,
    monto: int | float | str | Decimal,
    referencia: str,
    aplicaciones: list[dict],
    comentarios: str | None = None,
) -> int:
    """
    Registra un abono y sus aplicaciones dentro de una sola
    transacción SQL.

    aplicaciones:
    [
        {"obligacion_id": 4, "monto": 100000},
        {"obligacion_id": 5, "monto": 50000},
    ]

    Devuelve ID_ABONO.
    """

    try:
        date.fromisoformat(fecha)
    except ValueError as error:
        raise ValueError(
            "FECHA debe utilizar el formato YYYY-MM-DD"
        ) from error

    monto_abono = dinero(monto)

    if monto_abono <= 0:
        raise ValueError("El monto del abono debe ser positivo")

    if not aplicaciones:
        raise ValueError("El abono debe contener aplicaciones")

    aplicaciones_normalizadas: list[dict] = []
    total_por_obligacion: dict[int, Decimal] = {}

    for aplicacion in aplicaciones:
        obligacion_id = int(aplicacion["obligacion_id"])
        monto_aplicado = dinero(aplicacion["monto"])

        if monto_aplicado <= 0:
            raise ValueError(
                "Todos los montos aplicados deben ser positivos"
            )

        aplicaciones_normalizadas.append(
            {
                "obligacion_id": obligacion_id,
                "monto": monto_aplicado,
            }
        )

        total_por_obligacion[obligacion_id] = (
            total_por_obligacion.get(obligacion_id, Decimal("0.00"))
            + monto_aplicado
        )

    total_aplicado = sum(
        (
            aplicacion["monto"]
            for aplicacion in aplicaciones_normalizadas
        ),
        Decimal("0.00"),
    )

    if total_aplicado != monto_abono:
        raise ValueError(
            f"El abono es {monto_abono}, "
            f"pero sus aplicaciones suman {total_aplicado}"
        )

    with transaccion(ruta_bd) as conexion:

        saldos_actuales: dict[int, Decimal] = {}

        for obligacion_id, nueva_aplicacion in total_por_obligacion.items():

            fila = conexion.execute(
                """
                SELECT
                    D.MONTO
                    - COALESCE((
                        SELECT SUM(FA.MONTO_AMPARADO)
                        FROM tblFinAplicaciones AS FA
                        WHERE FA.ID_DPP = D.OBLIGACION_ID
                          AND FA.ACTIVO = 1
                    ), 0)
                    - COALESCE((
                        SELECT SUM(AA.MONTO)
                        FROM tblAplicacionesAbonos AS AA
                        WHERE AA.OBLIGACION_ID = D.OBLIGACION_ID
                          AND AA.ACTIVO = 1
                    ), 0) AS SALDO
                FROM tblDoctosXPagar AS D
                WHERE D.OBLIGACION_ID = ?
                  AND D.ACTIVO = 1
                """,
                (obligacion_id,),
            ).fetchone()

            if fila is None:
                raise ValueError(
                    f"La obligación {obligacion_id} no existe o no está activa"
                )

            saldo = dinero(fila["SALDO"])

            if nueva_aplicacion > saldo:
                raise ValueError(
                    f"La obligación {obligacion_id} tiene saldo {saldo}, "
                    f"pero se intentan aplicar {nueva_aplicacion}"
                )

            saldos_actuales[obligacion_id] = saldo

        cursor = conexion.execute(
            """
            INSERT INTO tblAbonos (
                FECHA,
                MONTO,
                REFERENCIA,
                ACTIVO,
                COMENTARIOS
            )
            VALUES (?, ?, ?, 1, ?)
            """,
            (
                fecha,
                float(monto_abono),
                referencia,
                comentarios,
            ),
        )

        id_abono = cursor.lastrowid

        for aplicacion in aplicaciones_normalizadas:
            conexion.execute(
                """
                INSERT INTO tblAplicacionesAbonos (
                    ABONO_ID,
                    OBLIGACION_ID,
                    MONTO,
                    ACTIVO,
                    COMENTARIOS
                )
                VALUES (?, ?, ?, 1, ?)
                """,
                (
                    id_abono,
                    aplicacion["obligacion_id"],
                    float(aplicacion["monto"]),
                    comentarios,
                ),
            )

        for obligacion_id, monto_nuevo in total_por_obligacion.items():
            saldo_final = saldos_actuales[obligacion_id] - monto_nuevo

            conexion.execute(
                """
                UPDATE tblDoctosXPagar
                SET PAGADO = ?
                WHERE OBLIGACION_ID = ?
                """,
                (
                    1 if saldo_final == Decimal("0.00") else 0,
                    obligacion_id,
                ),
            )

    return id_abono

File: frontend/src-tauri/src/abonos.rs
use crate::db;
use crate::obligation_state::aplicar_monto;

use rusqlite::{params, TransactionBehavior};

use serde::{Deserialize, Serialize};

use std::collections::HashMap;

#[derive(Debug, Deserialize)]
pub struct AplicacionAbonoEntrada {
    pub obligacion_id: i64,
    pub monto: String,
}

#[derive(Debug, Serialize)]
pub struct AbonoRegistrado {
    pub id_abono: i64,
    #[serde(serialize_with = "crate::money::serializar_centavos")]
    pub monto: i64,
    pub aplicaciones: usize,
}

#[tauri::command]
pub fn registrar_abono(
    fecha: String,
    monto: String,
    referencia: String,
    aplicaciones: Vec<AplicacionAbonoEntrada>,
    comentarios: Option<String>,
) -> Result<AbonoRegistrado, String> {
    crate::validation::validar_fecha_iso(&fecha, "FECHA")?;

    let monto_abono = crate::validation::dinero_a_centavos(&monto, "El monto del abono")?;

    if monto_abono <= 0 {
        return Err("El monto del abono debe ser positivo".to_string());
    }

    if aplicaciones.is_empty() {
        return Err("El abono debe contener aplicaciones".to_string());
    }

    let mut aplicaciones_normalizadas = Vec::with_capacity(aplicaciones.len());

    let mut total_por_obligacion: HashMap<i64, i64> = HashMap::new();

    let mut total_aplicado: i64 = 0;

    for aplicacion in aplicaciones {
        if aplicacion.obligacion_id <= 0 {
            return Err("OBLIGACION_ID debe ser positivo".to_string());
        }

        let monto_aplicado =
            crate::validation::dinero_a_centavos(&aplicacion.monto, "El monto aplicado")?;

        if monto_aplicado <= 0 {
            return Err("Todos los montos aplicados deben ser positivos".to_string());
        }

        total_aplicado = total_aplicado
            .checked_add(monto_aplicado)
            .ok_or_else(|| "La suma de aplicaciones excede el importe permitido".to_string())?;

        let acumulado = total_por_obligacion
            .entry(aplicacion.obligacion_id)
            .or_insert(0);

        *acumulado = acumulado.checked_add(monto_aplicado).ok_or_else(|| {
            format!(
                "Las aplicaciones de la obligación {} exceden el importe permitido",
                aplicacion.obligacion_id
            )
        })?;

        aplicaciones_normalizadas.push((aplicacion.obligacion_id, monto_aplicado));
    }

    if total_aplicado != monto_abono {
        return Err(format!(
            "El abono es {}, pero sus aplicaciones suman {}",
            crate::money::formatear_centavos(monto_abono),
            crate::money::formatear_centavos(total_aplicado),
        ));
    }

    let mut conexion = db::abrir_bd_escritura()?;

    let transaccion = conexion
        .transaction_with_behavior(TransactionBehavior::Immediate)
        .map_err(|error| format!("No fue posible iniciar la transacción: {error}"))?;

    let resultado = (|| -> Result<AbonoRegistrado, String> {
        for (obligacion_id, monto_aplicado) in &total_por_obligacion {
            aplicar_monto(&transaccion, *obligacion_id, *monto_aplicado)?;
        }

        transaccion
            .execute(
                r#"
                    INSERT INTO tblAbonos (
                        FECHA,
                        MONTO,
                        REFERENCIA,
                        ACTIVO,
                        COMENTARIOS
                    )
                    VALUES (?1, ?2, ?3, 1, ?4)
                    "#,
                params![fecha, monto_abono, referencia, comentarios,],
            )
            .map_err(|error| format!("No fue posible registrar el abono: {error}"))?;

        let id_abono = transaccion.last_insert_rowid();

        for (obligacion_id, monto_aplicado) in &aplicaciones_normalizadas {
            transaccion
                .execute(
                    r#"
                        INSERT INTO tblAplicacionesAbonos (
                            ABONO_ID,
                            OBLIGACION_ID,
                            MONTO,
                            ACTIVO,
                            COMENTARIOS
                        )
                        VALUES (?1, ?2, ?3, 1, ?4)
                        "#,
                    params![id_abono, obligacion_id, monto_aplicado, comentarios,],
                )
                .map_err(|error| {
                    format!(
                        "No fue posible aplicar el abono a la obligación {}: {}",
                        obligacion_id, error
                    )
                })?;
        }

        Ok(AbonoRegistrado {
            id_abono,
            monto: monto_abono,
            aplicaciones: aplicaciones_normalizadas.len(),
        })
    })();

    match resultado {
        Ok(abono) => {
            transaccion
                .commit()
                .map_err(|error| format!("No fue posible confirmar el abono: {error}"))?;

            Ok(abono)
        }

        Err(error) => {
            // Al salir sin commit, rusqlite ejecuta rollback.
            Err(error)
        }
    }
}
